Smarter "Collect images" filter

oexler opened the pull request:
Here are some changes to the collect images panel
- collect options depend on other rules
- show image count for the option
- hierarchical dates
- selecting "AND NOT" does not clear collection until something is selected
But since I don't have a large collection, I don't know whether it will be fast enough, as multiple tables are joined.

TurboGit commented:
Nice! I had started to work on the count for a collection but never finish, this is really a neat new feature. I've not look at the code yet but I found an issue. If you select a collection and then reset the module (click on the small circle on the right of the module) then the lightable is reset properly but not the content of the module itself (the previously selected collection is still there).

TurboGit commented on src/common/collection.c:
```diff
@@ -539,50 +547,9 @@ int dt_collection_get_nth(const dt_collection_t *collection, int nth)
-  g_free(query);
-
-  return list;
+  return dt_collection_get(collection,limit,TRUE);
```
spaces after colon.

TurboGit commented on src/common/collection.c:
```diff
@@ -515,6 +518,11 @@ GList *dt_collection_get_all(const dt_collection_t *collection, int limit)
 
+GList *dt_collection_get_all(const dt_collection_t *collection, int limit)
+{
+  return dt_collection_get(collection,limit,FALSE);
```
spaces after colons

TurboGit commented on src/libs/collect.c:
```diff
@@ -1219,13 +1231,13 @@ static void list_view(dt_lib_collect_rule_t *dr)
-                          "ORDER BY datetime_taken DESC",
-                   darktable.collection->where_ext);
-        break;
+//      case DT_COLLECTION_PROP_DAY:
```
you probably want to remove this commented out code.

TurboGit commented:
I would do the hierarchical view for date/time tag too. Maybe just on the date?

oexler commented:
Thank you for your input. 
I tried to implement your remarks.

TurboGit commented:
Works great now! The date-time is lot more useful as it is now.

A small issue, select a tag based on a color label. Remove the corresponding color label on one of the picture. The count on the collection module is not updated. A minor issue but if we can fix this it would be nice.

Also, If you select a collection AND NOT (exclude) those that have the tag exported (darktable|exported), exporting one of the picture will remove the picture from the lightable but the count won't be updated. Would be nice to fix that too.

oexler commented:
I tried a quick fix for the color labels, but noticed that it would need more changes.
As currently when you select a color label this filter is included in the darktable.collection->where_ext. So when I update the collection module it is limited to this color.
It is the same issue when you switch to a previously created rule.

When I have more time perhaps I try to see if it is possible to split the current rule from where_ext.

TurboGit commented on src/libs/collect.c:
```diff
@@ -1184,18 +1260,22 @@ static void list_view(dt_lib_collect_rule_t *dr)
           folder = dt_image_film_roll_name(folder);
         }
         gchar *value = (gchar *)sqlite3_column_text(stmt, 0);
+        int count = sqlite3_column_int(stmt, 2);
```
could be const

TurboGit commented on src/libs/collect.c:
```diff
@@ -875,6 +933,8 @@ static void tree_view(dt_lib_collect_rule_t *dr)
       char *name_folded = g_utf8_casefold(name, -1);
       gchar *collate_key = NULL;
 
+      int count = sqlite3_column_int(stmt, 2);
```
const?

TurboGit commented on src/libs/collect.c:
```diff
@@ -902,6 +964,7 @@ static void tree_view(dt_lib_collect_rule_t *dr)
     {
       name_key_tuple_t *tuple = (name_key_tuple_t *)names->data;
       char *name = tuple->name;
+      int count = tuple->count;
```
const?

TurboGit commented on src/libs/collect.c:
```diff
@@ -835,7 +879,10 @@ static void tree_view(dt_lib_collect_rule_t *dr)
   gboolean folders = (property == DT_COLLECTION_PROP_FOLDERS);
   gboolean tags = (property == DT_COLLECTION_PROP_TAG);
-  const char *format_separator = folders ? "%s" G_DIR_SEPARATOR_S : "%s|";
+  gboolean days = (property == DT_COLLECTION_PROP_DAY);
```
sounds like days, times could be set const too?

oexler commented:
I have changed where_ext from string to array of strings, so I can exclude the current rule from it.
Also the collect module should update every time the collection is updated.

TurboGit commented:
Looks good to me now. The two scenario I had tested are now working fine and the count is properly updated. Thanks!

oexler commented:
I have made a small change to make the behaviour of the list more consistent with the tree view.
The list now only gets filtered when something is typed in the entry box.
This way it is easier to switch between different entries.

TurboGit commented:
@oexler : The list now only gets filtered when something is typed in the entry box.

I really like this, indeed you can click on different collections quickly. One annoying thing is that when you click on a collection it always move the collection tree to the top. This is very annoying when you are low on the list, you need to scroll down again and again to select another collection. When double clicking on a collection the tree should not reset to top but kept as-is.

oexler commented:
The problem was, that every change to the collection triggered an update for the collect list.
I have added a boolean to prevent an update when an entry is activated.
Additionally I added a call to scroll to the selected entry when the collection is updated from somewhere else or when switching between rules.

oexler commented:
What are the next steps to get this pr merged? 
Should I rebase or merge master to incorporate the changes for local copies? There seems to be no conflict but I could probably also show the image count.

TurboGit commented:
@oexler: yes rebasing and adding the local copy support would be nice.

Best way to proceed next is to go to IRC and talk with the core devs for integration. I'll talk about this module on IRC tonight if you want.

oexler commented:
I have rebased the pr. normally I would squash the changes to one commit, but I think it is easier to follow, as it incorporates different changes.
It would be nice if you talk about this pr on IRC.

TurboGit commented:
@oexler : wonderful! Indeed better to not squash as it makes it easier to review I think. I have raised the awareness of this PR on IRC. It is nice to see some refresh on this area for darktable, the filter is quite better this was to me. thanks again.

houz commented on src/libs/collect.c:
```diff
@@ -1439,12 +1489,14 @@ static void row_activated(GtkTreeView *view, GtkTreePath *path, GtkTreeViewColum
   else
     update_view(d->rule + active); // we have to update visible items too
 
+  d->skipnextupdate = TRUE;
```
What is the reason for this flag? Is `dt_collection_update_query` emitting a signal that in turn causes this module to rebuild its list? If that's the reason, then instead of having this flag and risking to run into a race condition when other parts of dt might cause the signal to be emitted at the same time, just disconnect from the signal, call `dt_collection_update_query` and connect again (using `g_signal_handlers_block_by_func`).

oexler replied:
When the signal is raised by other parts at the same time, the module should rebuilt its list.
The flag is just there to prevent it once, when something is selected from the list.
If I were to block the signal, there would be no update from another part calling the update concurrently, if I understood correctly.

houz commented on src/libs/collect.c:
```diff
@@ -1034,153 +1111,158 @@ static void list_view(dt_lib_collect_rule_t *dr)
-        g_strlcpy(query, "SELECT DISTINCT filename, 1 FROM main.images ORDER BY filename", sizeof(query));
-        break;
-
-      case DT_COLLECTION_PROP_DAY:
```
`DT_COLLECTION_PROP_{DAY,TIME}` are no longer handled. Is the default case working for those?

houz commented on src/libs/collect.c:
```diff
@@ -1530,95 +1579,61 @@ static void menuitem_and(GtkMenuItem *menuitem, dt_lib_collect_rule_t *d)
-    dt_conf_set_int(confname, DT_LIB_COLLECT_MODE_OR);
+    dt_lib_collect_mode_t mode = 0;
+    const gchar *label = gtk_menu_item_get_label (menuitem);
+    if(strcmp(label, _("narrow down search")) == 0)
```
I don't like comparing translated strings. Instead you should store the mode in the menuitem, similar to https://github.com/darktable-org/darktable/blob/master/src/gui/presets.c#L723 and then retrieve it here (https://github.com/darktable-org/darktable/blob/master/src/gui/presets.c#L582) – just with int instead of char.

oexler commented on src/libs/collect.c:
```diff
@@ -965,13 +1038,14 @@ static void tree_view(dt_lib_collect_rule_t *dr)
             gtk_tree_store_set(GTK_TREE_STORE(model), &iter, DT_LIB_COLLECT_COL_TEXT, *token,
-                               DT_LIB_COLLECT_COL_PATH, pth2, DT_LIB_COLLECT_COL_VISIBLE, TRUE, -1);
+                               DT_LIB_COLLECT_COL_PATH, pth2, DT_LIB_COLLECT_COL_VISIBLE, TRUE,
+                               DT_LIB_COLLECT_COL_COUNT, (*(token + 1)?0:count), -1);
```
Because I only know the count for the full path and the common path parts were already added to the tree before.
I could also recursively set the count for the parents of the current leaf, but then this function would be even longer. The callback is doing exactly the same and I think by splitting the code is easier to understand.

oexler commented:
Are there any open questions or comments?

TurboGit commented:
Just wondering if it would be possible to make the collapsable tree used to display date and datetime for example as a separate widget? It would probably help implementing https://redmine.darktable.org/issues/9323. Does this makes sense here? What do you think?

TurboGit commented:
Any new on this? If my request does not makes sense I'd like to merge this at this stage, it will get used by more people.

houz commented:
There are still a few things open that I didn't like. I am a little short on time right now so I can't go through all the code and come up with better solutions myself.

oexler commented:
I don't know how much of the functionality is reusable in a separate widget.
Currently a GtkTreeView widget is used to display the tree. The tree_view function from the collect module gets the strings and splits it depending on the type of data.

TurboGit commented:
@houz, and you want to fix this when it is merged? or before?

houz commented:
Before. We have too much code that got merged prematurely and never fixed. @oexler came to IRC earlier and we had a very brief talk.

oexler commented:
I have removed my skipnextupdate flag and instead disconnect from the signal.
I have created dt_control_signal_(un)block_by_func(). I don't know whether a more generic (un)block_matched would be better.

oexler commented:
I think I have completed the remaining change request.

TurboGit commented on src/common/collection.c:
```diff
@@ -264,13 +267,41 @@ void dt_collection_set_query_flags(const dt_collection_t *collection, uint32_t f
+    const int mode = dt_conf_get_int(confname);
+    if (mode != 1)
+    {
+      for(int i = 0; collection->where_ext[i] != NULL; i++) {
```
the brace should be on its own line

houz commented:
I like the new way to work with collect. Thank you a lot. The numbers are a little distracting though. We'll see if they are too distracting in the long run.
